src/main/scala/fpinscalalib/FunctionalStateSection.scala

```diff
@@ -0,0 +1,466 @@
+    * combine them into one action that generates pairs of both `A` and `B`:
+    *
+    * {{{
+    *   def both[A,B](ra: Rand[A], rb: Rand[B]): Rand[(A,B)] = map2(ra, rb)((_, _))
+    * }}}
+    *
+    * We can use this to reimplement `intDouble` and `doubleInt` from exercise 6.3 more succinctly:
+    *
+    * {{{
+    *   val randIntDouble: Rand[(Int, Double)] = both(int, double)
+    *
+    *   val randDoubleInt: Rand[(Double, Int)] = both(double, int)
+    * }}}
+    *
+    * If we can combine two `RNG` transitions, we should be able to combine a whole list of them. Let's implement
+    * `sequence` for combining a `List` of transitions into a single transition:
+    *
+    * {{{
+    *   def sequence[A](fs: List[Rand[A]]): Rand[List[A]] =
+    *     fs.foldRight(unit(List[A]()))((f, acc) => map2(f, acc)(_ :: _))
+    * }}}
+    *
+    * We can use `sequence` to reimplement the `ints` function we wrote before. To do so, let's use the standard library
+    * function `List.fill(n)(x)` to make a list with `x` repeated `n` times:
+    *
+    * {{{
+    *   def _ints(count: Int): Rand[List[Int]] =
+    *     sequence(List.fill(count)(int))
+    * }}}
+    *
+    * Our next step is to implement `flatMap`, as it will allow us to change state operations:
+    *
+    * {{{
+    *   def flatMap[A,B](f: Rand[A])(g: A => Rand[B]): Rand[B] =
+    *     rng => {
+    *       val (a, r1) = f(rng)
+    *       g(a)(r1)
+    *     }
+    * }}}
+    *
+    * Let's use it to write `nonNegativeLessThan`, which generates an integer between `0` (inclusive) and `n`
+    * (exclusive):
+    */
+
+  def randomNonNegativeLessThan(res0: Int, res1: Int): Unit = {
+    def nonNegativeLessThan(n: Int): Rand[Int] = {
+      flatMap(nonNegativeInt) { i =>
+        val mod = i % n
+        if (i + (n - res0) - mod >= res1) unit(mod) else nonNegativeLessThan(n)
+      }
+    }
+
+    val (r1, rng1) = nonNegativeLessThan(10)(Simple(47))
+    val r2 = nonNegativeLessThan(10)(rng1)._1
+
+    r1 should be >= 0
+    r1 should be < 10
+    r2 should be >= 0
+    r2 should be < 10
+    r1 should not be r2
+  }
+
+  /**
+    * We can also rewrite `map` and `map2` in terms of `flatMap`. The fact that this is possible is what we’re referring
+    * to when we say that `flatMap` is more powerful than `map` and `map2`.
+    *
+    * {{{
+    *   def _map[A,B](s: Rand[A])(f: A => B): Rand[B] =
+    *     flatMap(s)(a => unit(f(a)))
+    *    def _map2[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] =
+    *     flatMap(ra)(a => map(rb)(b => f(a, b)))
+    * }}}
+    *
+    * As a final example, let's revisit our previously written functions to implement a function that roll a six-sided
+    * die:
+    */
+
+  def randomRollDie(res0: Int): Unit = {
+    def rollDie: Rand[Int] = map(nonNegativeLessThan(6))(_ + res0)
+
+    val (d1, r1) = rollDie(Simple(47))
+    val d2 = rollDie(r1)._1
```

Comment:
We should use some meaningful names and attempt to deconstruct all tuples instead of referring to them by index name